Set proxy from environment variables when initialize request session

Isotr0py opened the pull request:
**Related issue**: 
- #501 

**Motivation**
- The direct connection to Earthdata through our university's network is slow and unstable, which always causes corrupted downloads even the failed search.
- We have a proxy to download data from data providers like Earthdata. However, setting proxy for `earthaccess` is a little bit tricky refer to the example code in #501.

**Description**:
- This PR lets user initialize the request session with proxy, if they have proxy set in environment variables.

<!--
Replace this text, including the symbols above and below it, with descriptive text about
the change you are proposing. Please include a reference to any issues addressed or
resolved in that text, for example "resolves #1".
-->

<!--
IMPORTANT: As a contributor, we would like as much help as you can offer, but we only
expect you to complete the steps in the "PR draft checklist" below. Maintainers are
willing and ready to help pick it up from there!

Please start by opening this Pull Request as a "draft". You can do this by
clicking the arrow on the right side of the green "Create pull request" button. While
your pull request is in "draft" state, maintainers will assume the PR isn't ready for
their attention unless they are specifically summoned using GitHub's @ system.

Follow the draft checklist below to move the PR out of draft state. If you accidentally
created the PR as a non-draft, don't worry, you can still change it to a draft using the
"Convert to draft" button on  the right side panel under the "Reviewers" section.
-->

<details><summary>Pull Request (PR) draft checklist - click to expand</summary>

- [ ] Please review our
      [contributing documentation](https://earthaccess.readthedocs.io/en/latest/contributing/)
      before getting started.
- [ ] Ensure an issue exists representing the problem being solved in this PR.
- [ ] Populate a descriptive title. For example, instead of "Updated README.md", use a
      title such as "Add testing details to the contributor section of the README".
      Example PRs: [#763](https://github.com/nsidc/earthaccess/pull/763)
- [ ] Populate the body of the pull request with:
    - A clear description of the change you are proposing.
    - Links to any issues resolved by this PR with text in the PR description, for
      example `closes #1`. See
      [GitHub docs - Linking a pull request to an issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
- [ ] Update `CHANGELOG.md` with details about your change in a section titled
      `## Unreleased`. If such a section does not exist, please create one. Follow
      [Common Changelog](https://common-changelog.org/) for your additions.
      Example PRs: [#763](https://github.com/nsidc/earthaccess/pull/763)
- [ ] Update the documentation and/or the `README.md` with details of changes to the
      earthaccess interface, if any. Consider new environment variables, function names,
      decorators, etc.

Click the "Ready for review" button at the bottom of the "Conversation" tab in GitHub
once these requirements are fulfilled. Don't worry if you see any test failures in
GitHub at this point!

</details>

<details><summary>Pull Request (PR) merge checklist - click to expand</summary>

Please do your best to complete these requirements! If you need help with any of these
requirements, you can ping the `@nsidc/earthaccess-support` team in a comment and we
will help you out!

- [ ] Add unit tests for any new features.
- [ ] Apply formatting and linting autofixes. You can add a GitHub comment in this Pull
      Request containing "pre-commit.ci autofix" to automate this.
- [ ] Ensure all automated PR checks (seen at the bottom of the "conversation" tab) pass.
- [ ] Get at least one approving review.

</details>

chuckwondo reviewed (changes requested):
Thanks for the contribution, but we'll need to discuss whether or not this is an appropriate/safe way to address #501. As I've commented there, you might be able to address the issue by exporting the appropriate env var, as there are cases where they are automatically used, and thus no need to explicitly set them on the session. Specifically, `requests` uses `urllib`, which recognizes the proxy env vars.

For cases where the proxy env vars are _not_ used, that's because the session's `trust_env` property is set to `False`, indicating that the env should _not_ be trusted. Therefore, the code in this PR _ignores_ the `trust_env` property, which is perhaps not the most secure approach to solving this.

Isotr0py commented:
Thanks for informing! I didn't notice the `trust_env` is set to `False` for safety before. 

So how about adding a method like `earthaccess.set_proxy()` to expose proxy settings to user?

chuckwondo commented:
> Thanks for informing! I didn't notice the `trust_env` is set to `False` for safety before.

It's set to `False` in _some_ cases, which is why I suggest trying to set the proxy env vars and seeing if that works. If it does, great. If not, then you've hit a case where `trust_env` is set to `False`.
 
> So how about adding a method like `earthaccess.set_proxy()` to expose proxy settings to user?

The auth logic in this library has arguably become a bit "messy," so we'll need to make a thorough review of the code to determine a reasonable approach, which might very well be roughly as simple as your suggestion.

Isotr0py commented:
> It's set to False in some cases, which is why I suggest trying to set the proxy env vars and seeing if that works. If it does, great. If not, then you've hit a case where trust_env is set to False.

I have double checked the proxy env vars, and it didn't work. I guess the case is that I'm using `interactive` strategy to login in, which will create bearer tokens and blocking env vars here:
https://github.com/nsidc/earthaccess/blob/4aa122390a78280a0b1a3af67abac21e6bf280bb/earthaccess/auth.py#L200-L216

chuckwondo commented:
> > It's set to False in some cases, which is why I suggest trying to set the proxy env vars and seeing if that works. If it does, great. If not, then you've hit a case where trust_env is set to False.
> 
> I have double checked the proxy env vars, and it didn't work. I guess the case is that I'm using `interactive` strategy to login in, which will create bearer tokens and blocking env vars here:

Can you try creating a `~/.netrc` file with your Earthdata Login creds?

Isotr0py commented:
> Can you try creating a ~/.netrc file with your Earthdata Login creds?

I tried both `netrc` and `environment` strategies, and they all return `session.trust_env=False` with header contents bearer tokens. I wonder if there is an issue related to the auth logic.

**My test script**
```python3
import earthaccess

earthaccess.login(strategy="netrc")
session = earthaccess.get_requests_https_session()
print(session.headers)
print(session.trust_env)
``` 

**Outputs**
```text
{'User-Agent': 'earthaccess v0.10.0', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Authorization': 'Bearer <My_token>'}
False
```

chuckwondo commented:
> > Can you try creating a ~/.netrc file with your Earthdata Login creds?
> 
> I tried both `netrc` and `environment` strategies, and they all return `session.trust_env=False` with header contents bearer tokens. I wonder if there is an issue related to the auth logic.

Don't bother looking at that for the moment. That may be misleading. Instead, try to do the thing that currently isn't working for you. For example, if you have a script to do something with earthaccess that currently fails due to not making use of the proxy, and the script is named `foo.py`, run it like so to see if it works:

```plain
https_proxy=http://HOST:PORT python foo.py
```

where HOST and PORT are whatever is appropriate for your proxy.

I am able so run a local proxy on my laptop at http://localhost:8080 and I am able to login and download via earthaccess.

Isotr0py commented:
I have a simple download script:
```python3
import argparse
import earthaccess
import os


def download(date_start, date_end, shortname):
    earthaccess.login()
    results = earthaccess.search_data(
        short_name=shortname,
        cloud_hosted=True,
        temporal=(date_start, date_end),
    )
    session = earthaccess.get_requests_https_session()
    print("http_proxy:", os.environ["http_proxy"])
    print("https_proxy:", os.environ["https_proxy"])
    print(session.headers)
    print(session.trust_env)
    earthaccess.download(results, f"download/{shortname}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--shortname", type=str)
    parser.add_argument("--date-start", type=str)
    parser.add_argument("--date-end", type=str)
    args = parser.parse_args()
    download(args.date_start, args.date_end, args.shortname)
```

And I run with below command on Windows Powershell, however, the download has failed with corrupted files, while `http_proxy` and `https_proxy` did has been passed to env vars.
```powershell
$env:HTTP_PROXY="http://127.0.0.1:7890"; $env:HTTPS_PROXY="http://127.0.0.1:7890"; python .\test_scripts.py --shortname GOCI_L2_OC --date-start 2015-01-01 --date-end 2015-01-02
```
**Outputs**
```
http_proxy: http://127.0.0.1:7890
https_proxy: http://127.0.0.1:7890
{'User-Agent': 'earthaccess v0.10.0', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Authorization': 'Bearer <my token>'}
False

QUEUEING TASKS | : 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 16/16 [00:00<00:00, 557.52it/s]
PROCESSING TASKS | :   0%|                                                                                                                                                                         | 0/16 [00:00<?, ?it/s]Error while downloading the file G2015001001640.L2_COMS_OC.nc
Traceback (most recent call last):
  File "F:\GIS\earthaccess\.venv\Lib\site-packages\urllib3\response.py", line 748, in _error_catcher
    yield
  File "F:\GIS\earthaccess\.venv\Lib\site-packages\urllib3\response.py", line 894, in _raw_read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
urllib3.exceptions.IncompleteRead: IncompleteRead(17812224 bytes read, 185151892 more expected)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "F:\GIS\earthaccess\earthaccess\store.py", line 646, in _download_file
    shutil.copyfileobj(r.raw, f, length=1024 * 1024)
  File "D:\miniconda3\Lib\shutil.py", line 197, in copyfileobj
    buf = fsrc_read(length)
          ^^^^^^^^^^^^^^^^^
  File "F:\GIS\earthaccess\.venv\Lib\site-packages\urllib3\response.py", line 949, in read
    data = self._raw_read(amt)
           ^^^^^^^^^^^^^^^^^^^
  File "F:\GIS\earthaccess\.venv\Lib\site-packages\urllib3\response.py", line 872, in _raw_read
    with self._error_catcher():
  File "D:\miniconda3\Lib\contextlib.py", line 155, in __exit__
    self.gen.throw(typ, value, traceback)
  File "F:\GIS\earthaccess\.venv\Lib\site-packages\urllib3\response.py", line 772, in _error_catcher
    raise ProtocolError(arg, e) from e
urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(17812224 bytes read, 185151892 more expected)', IncompleteRead(17812224 bytes read, 185151892 more expected))
```

However, when I switched to the script provided in https://github.com/nsidc/earthaccess/issues/501#issuecomment-2023663133, it can download 16 complete files through proxy.

chuckwondo commented:
Thank you. That's all helpful information.

chuckwondo commented:
@Isotr0py, I believe this might suffice as a workaround for you for the moment, so please let me know if this works for you. I've tweaked your sample code from above.

The basic idea is that I'm setting the proxies on the session obtained from `Auth.get_session` (which is invoked on the call path from `earthaccess.get_requests_https_session`) _and_ caching that session, such that every subsequent call to `Auth.get_session` will use that same session. This is necessary because otherwise, every call to `Auth.get_session` creates and returns a new session object.

This might be a way around modifying earthaccess code directly, so let me know if it does the trick.

```python
import argparse
import os
from functools import cache
from typing import Callable
from typing_extensions import ParamSpec

import earthaccess
import requests


P = ParamSpec("P")


def set_proxies(f: Callable[P, requests.Session]) -> Callable[P, requests.Session]:
    def go(*args: P.args, **kwargs: P.kwargs) -> requests.Session:
        s = f(*args, **kwargs)
        s.proxies.update(
            {
                k: v
                for k in ("HTTP_PROXY", "HTTPS_PROXY")
                if (v := os.environ.get(k, os.environ.get(k.casefold())))
            }
        )

        return s

    return go


def download(date_start, date_end, shortname):
    earthaccess.login()
    results = earthaccess.search_data(
        short_name=shortname,
        cloud_hosted=True,
        temporal=(date_start, date_end),
    )
    auth: earthaccess.Auth = earthaccess.__store__.auth
    auth.get_session = cache(set_proxies(auth.get_session))
    session = earthaccess.get_requests_https_session()
    print("http_proxy:", os.environ.get("http_proxy"))
    print("https_proxy:", os.environ.get("https_proxy"))
    print(f"{session.proxies=}")
    print(session.headers)
    print(session.trust_env)
    earthaccess.download(results, f"download/{shortname}")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--shortname", type=str)
    parser.add_argument("--date-start", type=str)
    parser.add_argument("--date-end", type=str)
    args = parser.parse_args()
    download(args.date_start, args.date_end, args.shortname)
```

Isotr0py commented:
@chuckwondo The above script didn't work, because the `set_proxy` is using incorrect keys (`HTTP_PROXY` and `HTTPS_PROXY`) in proxy dict. 

In fact, this script worked well after I modified the `set_proxies` with the correct keys (`http` and `https`):
```python3
def set_proxies(f: Callable[P, requests.Session]) -> Callable[P, requests.Session]:
    def go(*args: P.args, **kwargs: P.kwargs) -> requests.Session:
        s = f(*args, **kwargs)
        # session use 'http' and `https` as proxies dict keys
        s.proxies.update(
            {
                k.split("_")[0].lower(): v
                for k in ("HTTP_PROXY", "HTTPS_PROXY")
                if (v := os.environ.get(k, os.environ.get(k.casefold())))
            }
        )

        return s

    return go
```

chuckwondo reviewed (changes requested):
@Isotr0py, the example I provided was for you to use _instead_ of this code change. I'm glad it works for you (with the small tweak you made to it).

We still do _not_ want to make this particular PR change internally because, as I mentioned earlier, this completely ignores cases where the session's `trust_env` property is set to `False`. We want to respect that setting and instead find a better way to give the caller more control over the session settings themselves.

Please close this PR, as this is a larger issue that the earthaccess team needs to discuss and find a solution for.

chuckwondo commented:
> @chuckwondo The above script didn't work, because the `set_proxy` is using incorrect keys (`HTTP_PROXY` and `HTTPS_PROXY`) in proxy dict.
> 
> In fact, this script worked well after I modified the `set_proxies` with the correct keys (`http` and `https`):
> 
> ```python
> def set_proxies(f: Callable[P, requests.Session]) -> Callable[P, requests.Session]:
>     def go(*args: P.args, **kwargs: P.kwargs) -> requests.Session:
>         s = f(*args, **kwargs)
>         # session use 'http' and `https` as proxies dict keys
>         s.proxies.update(
>             {
>                 k.split("_")[0].lower(): v
>                 for k in ("HTTP_PROXY", "HTTPS_PROXY")
>                 if (v := os.environ.get(k, os.environ.get(k.casefold())))
>             }
>         )
> 
>         return s
> 
>     return go
> ```
> 
> **Outputs**
> 
> ```
> http_proxy: http://127.0.0.1:7890
> https_proxy: http://127.0.0.1:7890
> session.proxies={'http': 'http://127.0.0.1:7890', 'https': 'http://127.0.0.1:7890'}
> {'User-Agent': 'earthaccess v0.10.0', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Authorization': 'Bearer <my token>'}
> False
> QUEUEING TASKS | : 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:00<00:00, 776.63it/s]
> PROCESSING TASKS | : 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [01:44<00:00, 13.09s/it]
> ```

Awesome! I'm glad my suggestion (with your minor tweak) worked for you!

Personally, I'd probably tweak it like so, but just a personal preference:

```python
def set_proxies(f: Callable[P, requests.Session]) -> Callable[P, requests.Session]:
    def go(*args: P.args, **kwargs: P.kwargs) -> requests.Session:
        s = f(*args, **kwargs)
        s.proxies.update(
            {
                scheme: v
                for scheme in ("http", "https")
                if (
                    v := os.environ.get(
                        k := f"{scheme}_proxy", os.environ.get(k.upper())
                    )
                )
            }
        )

        return s

    return go
```
